### PR TITLE
`dependents` and `dependencies` are scoped and more useful now

### LINKED ISSUE #2379: sort the output of dependents / dependencies
Ah, it looks like we need to use Alphabetical or whatever; I guess these are currently sorted by ... I'm not sure if it's suffix or somehow reference.
```
    81.  ##Bytes.at                    base.Bytes.at
    82.  ##Bytes.at                    external.paulcadman.unison_parser.base.Bytes.at
    83.  ##Bytes.at                    builtin.Bytes.at
    84.  ##List.at                     tmp._atln.List.at
    85.  ##List.at                     base.List.at
    86.  ##List.at                     external.paulcadman.unison_parser.base.List.at
    87.  ##List.at                     builtin.List.at
    88.  #uum1piclle                   base.Bytes.at!
    89.  #glts8qofgo                   external.unison_random_mersenne.v1._base_additions.Nat32.bit
    90.  #q8rms9hf34                   base.Random.impl.bytesFromNats
    91.  #f8v9prk5ij                   tmp.c
    92.  #5blc2jrop4                   base.Nat.choose
    93.  #ho3f6sc7fg                   external.paulcadman.unison_parser.base.Nat.choose
    94.  #77hkmnfup8                   external.unison_random_mersenne.v1._base_additions.Nat32.clearBit
    95.  #49dfoqk8ri                   external.unison_random_mersenne.v1.random.mersenne.collect
    96.  #oeu30vnok3                   base.Bag.internal.compare
    97.  #oeu30vnok3                   external.paulcadman.unison_parser.base.Bag.internal.compare
    98.  ##Nat.complement              tmp._atln.Nat.complement
    99.  ##Nat.complement              base.Nat.complement
    100. ##Nat.complement              external.paulcadman.unison_parser.base.Nat.complement
    101. ##Nat.complement              builtin.Nat.complement
[less] Use space/arrow keys to navigate, or 'q' to return to ucm
```

### LINKED ISSUE #3803: `dependents` and `dependencies` should be scoped
The `dependents` and `dependencies` functions are currently outliers because they operate in the global namespace instead of the current namespace. Instead they should only find dependents/dependencies in the _current_ namespace. Or, if projects are supported before this ticket is addressed then they should probably be scoped to a project instead of a namespace.

This was mentioned in #3403 but I didn't create a proper issue for it.

As a bonus, applying these commands with a smaller scope will probably make them faster.

### PR DESCRIPTION
**Pull request by pchiusano**
## Overview

Closes #2379, Closes #3803 

`dependencies` and `dependents` commands previously showed dependencies globally and the output wasn't very pretty.

<details><summary>☢️ HORRIFYING OLD OUTPUT, AVERT YOUR EYES ☢️ </summary>
image
<pre>
    81.  ##Bytes.at                    base.Bytes.at
    82.  ##Bytes.at                    external.paulcadman.unison_parser.base.Bytes.at
    83.  ##Bytes.at                    builtin.Bytes.at
    84.  ##List.at                     tmp._atln.List.at
    85.  ##List.at                     base.List.at
    86.  ##List.at                     external.paulcadman.unison_parser.base.List.at
    87.  ##List.at                     builtin.List.at
    88.  #uum1piclle                   base.Bytes.at!
    89.  #glts8qofgo                   external.unison_random_mersenne.v1._base_additions.Nat32.bit
    90.  #q8rms9hf34                   base.Random.impl.bytesFromNats
    91.  #f8v9prk5ij                   tmp.c
    92.  #5blc2jrop4                   base.Nat.choose
    93.  #ho3f6sc7fg                   external.paulcadman.unison_parser.base.Nat.choose
    94.  #77hkmnfup8                   external.unison_random_mersenne.v1._base_additions.Nat32.clearBit
    95.  #49dfoqk8ri                   external.unison_random_mersenne.v1.random.mersenne.collect
    96.  #oeu30vnok3                   base.Bag.internal.compare
    97.  #oeu30vnok3                   external.paulcadman.unison_parser.base.Bag.internal.compare
    98.  ##Nat.complement              tmp._atln.Nat.complement
    99.  ##Nat.complement              base.Nat.complement
    100. ##Nat.complement              external.paulcadman.unison_parser.base.Nat.complement
    101. ##Nat.complement              builtin.Nat.complement
</pre>
</details>

Now, both commands show results from the current namespace. `dependents` also skips things in `lib` (you can `cd` into `lib` if you'd like to see dependents within there). Output is now sorted by name and looks nice:

image

Both `dependents` and `dependencies` also use the same code path for rendering, and look identical: 

image

## Implementation notes

Nothing interesting.

## Interesting/controversial decisions

* `dependents` skips things in `lib`, and this is what you want 95% of the time - you're looking for usages of a thing in your project, not in the dependencies of your project. You can always `cd lib` and then `dependents` in there if you want. (`dependencies` shows ALL dependencies though, regardless of whether they have names in the current namespace)
* If the user types `dependencies Foo`, and `Foo` is both a type and a term, or `Foo` is ambiguous, the output is union'd together (and it mentions the list of terms and types it is showing results for). The user can always repeat the command with a more specific name or a hash qualified name to disambiguate.

## Test coverage

Decent coverage from existing transcript. I think it's fine.

## Loose ends

None. However, I noticed #3992. `dependents` will be a lot more useful once that's fixed.

### CODE REVIEW

**Comment by ceedubs** [comment]
I'll leave the code reviewing to others, but 1000x YES for the change!

File: unison-cli/src/Unison/Codebase/Editor/HandleInput.hs [hunk 1]
```diff
@@ -1773,27 +1784,21 @@ handleDependents hq = do
...
-                g :: HQ'.HashQualified Name -> (Reference, Maybe Name)
-                g hqName =
-                  (reference, Just (HQ'.toName hqName))
-    #numberedArgs .= map (Text.unpack . Reference.toText . fst) results
-    Cli.respond (ListDependents hqLength ld results)
+    let -- True is type names, False is term names
+        results :: [(Bool, HQ.HashQualified Name, Reference)]
+        results = do
+          r <- Set.toList dependents
+          Just (isTerm, hq) <- [(True,) <$> PPE.terms fqppe (Referent.Ref r), (False,) <$> PPE.types fqppe r]
```
Comment (aryairani) [inline unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1791]: Haskell question, does this `Just (isTerm, hq)` end up being a filter? or a potential pattern match failure? or guaranteed to match for some reason?

Comment (pchiusano) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1791]: It calls `fail` for the monad in question, which for list returns `[]`.

```
» ghci
GHCi, version 8.10.4: https://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /Users/pchiusano/.ghci
λ: do { Just a <- [Nothing, Nothing, Just 4]; pure a }
[4]
```

Comment (pchiusano) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1791]:
```
λ: fail "hi" :: [Int]
[]
```

Comment (aryairani) [reply unison-cli/src/Unison/Codebase/Editor/HandleInput.hs:1791]: Ah, I see.

**Review by aryairani — APPROVED** [review]
Great, thanks for putting this together

**Pull request merged**